src/subject_name/verify.rs

```diff
@@ -141,137 +141,122 @@ fn check_presented_id_conforms_to_constraints(
-                //
-                // Rejecting everything technically confirms to RFC5280:
-                //
-                //   "If a name constraints extension that is marked as critical imposes constraints
-                //    on a particular name form, and an instance of that name form appears in the
-                //    subject field or subjectAltName extension of a subsequent certificate, then
-                //    the application MUST either process the constraint or _reject the certificate_."
-                //
-                // TODO: rustls/webpki#19
-                //
-                // Rejection is achieved by not matching any PermittedSubtrees, and matching all
-                // ExcludedSubtrees.
-                match subtrees {
-                    Subtrees::PermittedSubtrees => false,
-                    Subtrees::ExcludedSubtrees => true,
-                },
-            ),
-
-            (GeneralName::IpAddress(name), GeneralName::IpAddress(base)) => {
-                ip_address::presented_id_matches_constraint(name, base)
-            }
+                // RFC 4280 says "If a name constraints extension that is marked as
+                // critical imposes constraints on a particular name form, and an
+                // instance of that name form appears in the subject field or
+                // subjectAltName extension of a subsequent certificate, then the
+                // application MUST either process the constraint or reject the
+                // certificate." Later, the CABForum agreed to support non-critical
+                // constraints, so it is important to reject the cert without
+                // considering whether the name constraint it critical.
+                (GeneralName::Unsupported(name_tag), GeneralName::Unsupported(base_tag))
+                    if name_tag == base_tag =>
+                {
+                    Err(Error::NameConstraintViolation)
+                }
 
-            // RFC 4280 says "If a name constraints extension that is marked as
-            // critical imposes constraints on a particular name form, and an
-            // instance of that name form appears in the subject field or
-            // subjectAltName extension of a subsequent certificate, then the
-            // application MUST either process the constraint or reject the
-            // certificate." Later, the CABForum agreed to support non-critical
-            // constraints, so it is important to reject the cert without
-            // considering whether the name constraint it critical.
-            (GeneralName::Unsupported(name_tag), GeneralName::Unsupported(base_tag))
-                if name_tag == base_tag =>
-            {
-                Err(Error::NameConstraintViolation)
-            }
+                _ => {
+                    // mismatch between constraint and name types; continue with current
+                    // name and next constraint
+                    continue;
```

Comment:
feels a bit less readable because we have this `continue` in a two-deep loop (though the comment gives a hint of which loop it applies to if the reader missed the inner loop), and the loop bodies are quite long now - with load-bearing code that has to run after the inner loop completes. perhaps hoist `fn general_subtree` out of the loop body? or drop the previous `DirectoryName` commit?

Reply:
not a hard objection either way, it's a bit subjective.

Reply:
I hoisted `general_subtree()` out of the loop body, hopefully that helps a bit.

I think there's more that can be done here, but probably still makes sense to merge this first.